src/features/virtualDocumentTracker.ts

```diff
@@ -0,0 +1,91 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import {workspace, TextDocument} from 'vscode';
+import {OmniSharpServer} from '../omnisharp/server';
+import * as serverUtils from '../omnisharp/utils';
+import { FileChangeType } from '../omnisharp/protocol';
+import { IDisposable } from '../Disposable';
+import CompositeDisposable from '../CompositeDisposable';
+
+function trackCurrentVirtualDocuments(server: OmniSharpServer) {
+    let registration = server.onProjectAdded(() => {
+        for (let i = 0; i < workspace.textDocuments.length; i++) {
+            let document = workspace.textDocuments[i];
+    
+            if (!shouldIgnoreDocument(document, server)) {
+                openVirtualDocument(document, server);
+            }
+        }
+
+        registration.dispose();
+    });
+}
+
+function trackFutureVirtualDocuments(server: OmniSharpServer): IDisposable {
+    let onTextDocumentOpen = workspace.onDidOpenTextDocument(document => {
+        if (shouldIgnoreDocument(document, server)) {
+            return;
+        }
+
+        openVirtualDocument(document, server);
+    });
+
+    let onTextDocumentClose = workspace.onDidCloseTextDocument(document => {
+        if (shouldIgnoreDocument(document, server)) {
+            return;
+        }
+
+        closeVirtualDocument(document, server);
+    });
+    
+    // We already track text document changes for virtual documents in our change forwarder.
+    return new CompositeDisposable(
+        onTextDocumentOpen,
+        onTextDocumentClose);
+}
+
+function shouldIgnoreDocument(document: TextDocument, server: OmniSharpServer): boolean {
+    if (document.uri.scheme === 'file' || document.languageId !== 'csharp') {
+        // We're only interested in non-physical CSharp documents.
+        return true;
+    }
+
+    if (!server.isRunning()) {
+        return true;
+    }
+
+    return false;
+}
+
+function openVirtualDocument(document: TextDocument, server: OmniSharpServer) {
+    let req = { FileName: document.uri.path, changeType: FileChangeType.Create };
+    serverUtils.filesChanged(server, [req])
```

Comment:
filesChanged and updateBuffer are both async/promise returning and should be awaited

Reply:
I thought so as well but that's not done in the `changeForwarding` class. Was trying to be consistent. Is it a mistake that it's not awaited there?

Reply:
In theory, maybe it doesn't matter because your requests get put into the queue in the submission order.

Reply:
@NTaylorMullen Recently we are trying to move to using async/await rather than .then/catch for promises. The `changeForwarding` still has the old pattern, but for this class we should prefer try/catch and async/await